relays/ethereum/src/ethereum_deploy_contract.rs

```diff
@@ -0,0 +1,136 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::ethereum_client::{self, EthereumConnectionParams, EthereumSigningParams};
+use crate::substrate_client::{self, SubstrateConnectionParams};
+use crate::substrate_types::{Hash as SubstrateHash, Header as SubstrateHeader};
+use codec::{Decode, Encode};
+use num_traits::Zero;
+
+/// Ethereum synchronization parameters.
+#[derive(Debug)]
+pub struct EthereumDeployContractParams {
+	/// Ethereum connection params.
+	pub eth: EthereumConnectionParams,
+	/// Ethereum signing params.
+	pub eth_sign: EthereumSigningParams,
+	/// Ethereum contract bytecode.
+	pub eth_contract_code: Vec<u8>,
+	/// Substrate connection params.
+	pub sub: SubstrateConnectionParams,
+	/// Initial authorities set id.
+	pub sub_initial_authorities_set_id: Option<u64>,
+	/// Initial authorities set.
+	pub sub_initial_authorities_set: Option<Vec<u8>>,
+	/// Initial header.
+	pub sub_initial_header: Option<Vec<u8>>,
+}
+
+impl Default for EthereumDeployContractParams {
+	fn default() -> Self {
+		EthereumDeployContractParams {
+			eth: Default::default(),
+			eth_sign: Default::default(),
+			// compiler: 0.6.6+commit.6c089d02 + optimization
+			eth_contract_code: hex::decode(include_str!("../res/substrate-bridge-bytecode.hex"))
+				.expect("code is hardcoded, thus valid; qed"),
+			sub: Default::default(),
+			sub_initial_authorities_set_id: None,
+			sub_initial_authorities_set: None,
+			sub_initial_header: None,
+		}
+	}
+}
+
+/// Deploy Bridge contract on Ethereum chain.
+pub fn run(params: EthereumDeployContractParams) {
+	let mut local_pool = futures::executor::LocalPool::new();
+
+	let result = local_pool.run_until(async move {
+		let eth_client = ethereum_client::client(params.eth);
+		let sub_client = substrate_client::client(params.sub);
+
+		let (sub_client, initial_header) = prepare_initial_header(sub_client, params.sub_initial_header).await;
+		let (initial_header_hash, initial_header) = initial_header?;
+		let initial_set_id = params.sub_initial_authorities_set_id.unwrap_or(0);
+		let (_, initial_set) = prepare_initial_authorities_set(
+			sub_client,
+			initial_header_hash,
+			params.sub_initial_authorities_set,
+		).await;
+		let initial_set = initial_set?;
+
+		log::info!(
+			target: "bridge",
+			"Deploying Ethereum contract.\r\n\tInitial header: {}\r\n\tInitial authorities set ID: {}\r\n\tInitial authorities set: {}",
+			hex::encode(&initial_header),
```

Comment:
It'd be worth displaying decoded (human-readable) values as well.

Reply:
Added header logging. Authorities are opaque to relay.